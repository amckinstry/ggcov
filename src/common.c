/*
 * ggcov - A GTK frontend for exploring gcov coverage data
 * Copyright (c) 2001-2005 Greg Banks <gnb@users.sourceforge.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <sys/types.h>
#include <sys/time.h>
#include "common.h"

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

extern char *argv0;

void
fatal(const char *fmt, ...)
{
    va_list args;

    va_start(args, fmt);
    fprintf(stderr, "%s: ", argv0);
    vfprintf(stderr, fmt, args);
    va_end(args);

    exit(1);
}

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

extern char *argv0;

static void oom(void) __attribute__((noreturn));

/* Do write() but handle short writes */
static int
do_write(int fd, const char *buf, int len)
{
    int remain = len;

    while (remain > 0)
    {
	int n = write(fd, buf, remain);
	if (n < 0)
	    return -1;
	buf += n;
	remain -= n;
    }
    return len;
}

static void
oom(void)
{
    static const char message[] = ": no memory, exiting\n";

    if (do_write(2, argv0, strlen(argv0)) < 0)
	exit(1);
    if (do_write(2, message, sizeof(message)-1) < 0)
	exit(1);
    exit(1);
}

void *
gnb_xmalloc(size_t sz)
{
    void *x;

    if ((x = g_malloc0(sz)) == 0)
	oom();

    return x;
}

void *
operator new(size_t sz)
{
    return gnb_xmalloc(sz);
}

void
operator delete(void *p)
{
    if (p != 0)     /* delete 0 is explicitly allowed */
	free(p);
}

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

int
u32cmp(uint32_t ul1, uint32_t ul2)
{
    if (ul1 > ul2)
	return 1;
    if (ul1 < ul2)
	return -1;
    return 0;
}

int
u64cmp(uint64_t ull1, uint64_t ull2)
{
    if (ull1 > ull2)
	return 1;
    if (ull1 < ull2)
	return -1;
    return 0;
}

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

unsigned long debug = 0;


static struct
{
    const char *name;
    unsigned int mask;
} const debug_tokens[] = {
#define F(feature)              (1UL<<(feature))
#define U(feature)              (F((feature)+1)-1)
#define R(feat1, feat2)         (U(feat2) & ~U(feat1))
{"verbose",     1},

{"files",       F(D_FILES)},
{"cpp",         F(D_CPP)},
{"bb",          F(D_BB)},
{"bbg",         F(D_BBG)},
{"da",          F(D_DA)},
{"io",          F(D_IO)},
{"cgraph",      F(D_CGRAPH)},
{"suppress",    F(D_SUPPRESS)},
{"solve",       F(D_SOLVE)},
{"stabs",       F(D_STABS)},
{"dwarf",       F(D_DWARF)},
{"elf",         F(D_ELF)},
{"dump",        F(D_DUMP)},
{"scene",       F(D_SCENE)},
{"report",      F(D_REPORT)},
{"dcallgraph",  F(D_DCALLGRAPH)},
{"dlego",       F(D_DLEGO)},

{"uicore",      F(D_UICORE)},
{"summarywin",  F(D_SUMMARYWIN)},
{"sourcewin",   F(D_SOURCEWIN)},
{"prefswin",    F(D_PREFSWIN)},
{"callswin",    F(D_CALLSWIN)},
{"funcswin",    F(D_FUNCSWIN)},
{"fileswin",    F(D_FILESWIN)},
{"graphwin",    F(D_GRAPHWIN)},
{"diagwin",     F(D_DIAGWIN)},
{"reportwin",   F(D_REPORTWIN)},

{"web",         F(D_WEB)},

#define SPECIALS_START     (const char *)1
{SPECIALS_START, 0},

{"cov",         R(D_FILES,D_DUMP)},
{"ui",          R(D_UICORE,D_DIAGWIN)},
{"none",        0},
{"all",         ~1U /*"all" doesn't get you "verbose"*/},
{0, 0}
#undef R
#undef U
#undef F
};

void
debug_set(const char *str)
{
    char *buf, *buf2, *x;
    int i;
    int dirn;

    buf = buf2 = g_strdup(str);
    debug = 0;

    while ((x = strtok(buf2, " \t\n\r,")) != 0)
    {
	buf2 = 0;

	if (*x == '+')
	{
	    dirn = 1;
	    x++;
	}
	else if (*x == '-')
	{
	    dirn = -1;
	    x++;
	}
	else
	    dirn = 1;

	for (i = 0 ; debug_tokens[i].name ; i++)
	{
	    if (debug_tokens[i].name == SPECIALS_START)
		continue;
	    if (!strcmp(debug_tokens[i].name, x))
	    {
		if (dirn == 1)
		    debug |= debug_tokens[i].mask;
		else
		    debug &= ~debug_tokens[i].mask;
		break;
	    }
	}
	if (!debug_tokens[i].name)
	    fprintf(stderr, "Unknown debug token \"%s\", ignoring", x);
    }

    g_free(buf);
}

char *
debug_enabled_tokens(void)
{
    int i;
    int len = 0;
    char *str, *p;

    for (i = 0 ; debug_tokens[i].name != SPECIALS_START ; i++)
    {
	if (!(debug & debug_tokens[i].mask))
	    continue;
	len += strlen(debug_tokens[i].name)+1;
    }

    p = str = (char *)gnb_xmalloc(len);

    for (i = 0 ; debug_tokens[i].name != SPECIALS_START ; i++)
    {
	if (!(debug & debug_tokens[i].mask))
	    continue;
	if (p != str)
	    *p++ = ',';
	strcpy(p, debug_tokens[i].name);
	p += strlen(p);
    }

    return str;
}

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

void timing_impl(const char *func, unsigned int line, const char *detail)
{
    static FILE *fp;
    static struct timeval prev;
    struct timeval now;
    static const char filename[] = "timing.dat";

    gettimeofday(&now, NULL);
    if (prev.tv_sec)
    {
	struct timeval d;
	timersub(&now, &prev, &d);
	if (!fp)
	{
	    fp = fopen(filename, "w");
	    if (!fp)
	    {
		perror(filename);
		exit(1);
	    }
	}
	fprintf(fp, "%s:%u %lu.%06lu %s\n",
		func, line,
		(unsigned long)d.tv_sec, (unsigned long)d.tv_usec,
		(detail ? detail : ""));
	fflush(fp);
    }
    prev = now;
}

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/
/*END*/
