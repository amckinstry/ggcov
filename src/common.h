/*
 * ggcov - A GTK frontend for exploring gcov coverage data
 * Copyright (c) 2001 Greg Banks <gnb@alphalink.com.au>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef _common_h_
#define _common_h_ 1

/* Include autoconf defines */
#include <config.h>

#if HAVE_STDINT_H
/*
 * Glibc <stdint.h> says:
 * The ISO C 9X standard specifies that in C++ implementations these
 * macros [UINT64_MAX et al] should only be defined if explicitly requested.
 */
#define __STDC_LIMIT_MACROS 1
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <malloc.h>
#include <string.h>
#include <assert.h>
#include <sys/stat.h>
#include <errno.h>
#include <ctype.h>
#if HAVE_STDINT_H
#include <stdint.h>
#endif

#include <glib.h>

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

#define strassign(v, s) \
    do { \
	if ((v) != 0) \
	    g_free((v)); \
	(v) = g_strdup((s)); \
    } while(0)
    
#define strdelete(v) \
    do { \
    	if ((v) != 0) \
	    g_free((v)); \
	(v) = 0; \
    } while(0)

#define strnullnorm(v) \
    do { \
	if ((v) != 0 && *(v) == '\0') \
	{ \
	    g_free((v)); \
	    (v) = 0; \
	} \
    } while(0)

#define listdelete(v,type,dtor) \
    do { \
	while ((v) != 0) \
	{ \
    	    dtor((type *)(v)->data); \
    	    (v) = g_list_remove_link((v), (v)); \
	} \
    } while(0)

#define listclear(v) \
    do { \
	while ((v) != 0) \
	{ \
    	    (v) = g_list_remove_link((v), (v)); \
	} \
    } while(0)

#define boolassign(bv, s) \
    (bv) = strbool((s), (bv))

#define boolstr(b)  	((b) ? "true" : "false")

#define safestr(s)  ((s) == 0 ? "" : (s))

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

#if defined(__ELF__) && defined(__GNUC__)
#define _CVSID(u,s) __asm__ (".ident \"" s "\"" )
#else
#define _CVSID(u,s) static const char * const __cvsid##u[2] = {(s), (const char*)__cvsid}
#endif
#define CVSID(s) _CVSID(,s)

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

extern void fatal(const char *fmt, ...)
#ifdef __GNUC__
__attribute__ (( noreturn ))
__attribute__ (( format(printf,1,2) ))
#endif
;

extern void *gnb_xmalloc(size_t sz);
#ifdef __cplusplus
/*
 * Operators new and delete operators are matched to avoid
 * complaints from valgrind and Purify about non-portable
 * mixing of malloc and operator delete.  Other than that,
 * the overloaded global operator delete provides no value.
 */
void *operator new(size_t sz);
void operator delete(void *);
#else
#define new(ty)     	((ty *)gnb_xmalloc(sizeof(ty)))
#define delete(p)     	free((p))
#endif

extern int ulcmp(unsigned long ul1, unsigned long ul2);
extern int ullcmp(unsigned long long ull1, unsigned long long ull2);

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

#ifdef _
#undef _
#endif
#define _(s)	s

#ifdef N_
#undef N_
#endif
#define N_(s)	s

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/


#endif /* _common_h_ */
