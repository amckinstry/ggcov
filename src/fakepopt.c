/*
 * ggcov - A GTK frontend for exploring gcov coverage data
 * Copyright (c) 2003 Greg Banks <gnb@users.sourceforge.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "fakepopt.h"

#ifndef HAVE_LIBPOPT

CVSID("$Id: fakepopt.c,v 1.5 2010-05-09 05:37:15 gnb Exp $");

/*
 * Simulate enough of the popt interface so that I only have
 * to write and maintain a single argument parsing routine for
 * all builds, e.g. gtk1.2, gtk2+popt, gtk2-popt.
 */

#ifdef __cplusplus
extern "C" {
#endif

struct _poptContext
{
    int argc;
    const char **argv;
    int argi;
    const char *current;

    int nfiles;
    const char **files;
    int filei;
    gboolean file_mode;

    char *other_option_help;

    const struct poptOption *options;
};

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

poptContext
poptGetContext(
    const char *ignored,
    int argc,
    const char **argv,
    const struct poptOption *options,
    int flags)
{
    poptContext con;

    con = g_new0(struct _poptContext, 1);

    con->argc = argc;
    con->argv = argv;
    con->argi = 0;  /* points at last argument visited */

    con->nfiles = 0;
    con->files = g_new0(const char *, con->argc);
    con->filei = 0;
    con->file_mode = FALSE;

    con->options = options;

    return con;
}

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

void
poptSetOtherOptionHelp(poptContext con, const char *help)
{
    con->other_option_help = g_strdup(help);
}

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

static const struct poptOption *
find_long_option(const struct poptOption *opt, const char *name, const char **valp)
{
    const struct poptOption *o;
    int len;

    for ( ; opt->long_name || opt->short_name ; opt++)
    {
	switch (opt->type)
	{
	case POPT_ARG_INCLUDE_TABLE:
	    o = find_long_option((const struct poptOption *)opt->value_ptr, name, valp);
	    if (o != 0)
		return o;
	    break;
	default:
	    len = strlen(opt->long_name);
	    if (!strncmp(name, opt->long_name, len) && name[len] == '=')
	    {
		*valp = name+len+1;
		return opt;
	    }
	    /* fall through */
	case POPT_ARG_NONE:
	    if (!strcmp(name, opt->long_name))
		return opt;
	    break;
	}
    }
    return 0;
}

static const struct poptOption *
find_short_option(const struct poptOption *opt, char name)
{
    const struct poptOption *o;

    for ( ; opt->long_name || opt->short_name ; opt++)
    {
	switch (opt->type)
	{
	case POPT_ARG_INCLUDE_TABLE:
	    o = find_short_option((const struct poptOption *)opt->value_ptr, name);
	    if (o != 0)
		return o;
	    break;
	default:
	    if (name == opt->short_name)
		return opt;
	    break;
	}
    }
    return 0;
}

static int
handle_option(poptContext con, const struct poptOption *opt, const char *val)
{
    if (opt == 0)
	return -2;  /* unknown option */
    assert(opt->value_ptr != 0);
    assert(opt->value == 0);
    switch (opt->type)
    {
    case POPT_ARG_NONE:
	assert(val == 0);
	*(int *)opt->value_ptr = TRUE;
	break;
    case POPT_ARG_STRING:
	if (val == 0)
	{
	    val = con->argv[++con->argi];
	    if (val == 0)
		return -3; /* no value */
	}
	*(const char **)opt->value_ptr = val;
	break;
    }
    return opt->value;
}

int
poptGetNextOpt(poptContext con)
{
    const char *arg;
    int rc;
    const char *val;
    const struct poptOption *opt;

    for (;;)
    {
	arg = con->argv[++con->argi];
	if (arg == 0)
	    return -1;  /* end of arguments */
	con->current = arg;

	if (con->file_mode || arg[0] != '-')
	{
	    con->files[con->nfiles++] = arg;
	}
	else if (arg[0] == '-' && arg[1] == '-')
	{
	    if (arg[2] == '\0')
	    {
		/* handle '--' */
		con->file_mode = TRUE;
		continue;
	    }
	    val = 0;
	    opt = find_long_option(con->options, arg+2, &val);
	    if ((rc = handle_option(con, opt, val)))
		return rc;
	}
	else
	{
	    for (arg++ ; *arg ; arg++)
	    {
		if ((rc = handle_option(con, find_short_option(con->options, *arg), 0)))
		    return rc;
	    }
	}
    }
    /*NOTREACHED*/
    return 0;
}

const char *
poptGetArg(poptContext con)
{
    if (con->filei >= con->nfiles)
	return 0;
    return con->files[con->filei++];
}

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/

const char *
poptStrerror(int rc)
{
    switch (rc)
    {
    case -2: return "bad option";
    case -3: return "no value";
    default: return "no error";
    }
}

const char *
poptBadOption(poptContext con, int ignored)
{
    return con->current;
}

void
poptFreeContext(poptContext con)
{
    if (con->other_option_help != 0)
	free(con->other_option_help);
    free(con->files);
    free(con);
}

#ifdef __cplusplus
}
#endif

#endif /* !HAVE_LIBPOPT */

/*-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-*/
/*END*/
