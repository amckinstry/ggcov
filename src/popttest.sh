#!/bin/sh

while read args ; do
    echo "===src/popttest $args"
    src/popttest $args 2>&1
done <<EOF

-b
--bool
-b -b
-b -S
--bool --spool
-bS
-s
-s FOO
--string FOO
--string=FOO
-b -S -s FOO
-bS -s FOO
-bSSb -s FOO
-bSSb -s FOO -R BAR
-bSSb -sR FOO BAR
--string FOO --rope BAR
--fmeh
-z
-R
foo
foo bar
-b foo bar
-b -- foo
-b -- -R foo
EOF
