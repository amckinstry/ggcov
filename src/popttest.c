#include "common.h"
#undef HAVE_LIBPOPT
#include "fakepopt.h"
#include "fakepopt.c"

static int boolopt = 0;
static int spoolopt = 0;
const char *stringopt = "before";
const char *ropeopt = "BEFORE";
GList *files;

static const struct poptOption included_options[] =
{
    {
    	"rope",	    	    	    	    	/* longname */
	'R',  	    	    	    	    	/* shortname */
	POPT_ARG_STRING,  	    	    	/* argInfo */
	&ropeopt,     	    	    	    	/* arg */
	0,  	    	    	    	    	/* val 0=don't return */
	"he's ropable",	    	    	    	/* descrip */
	0	    	    	    	    	/* argDescrip */
    },
    {
    	"spool",	    	    	    	/* longname */
	'S',  	    	    	    	    	/* shortname */
	POPT_ARG_NONE,  	    	    	/* argInfo */
	&spoolopt,     	    	    	    	/* arg */
	0,  	    	    	    	    	/* val 0=don't return */
	"spoolin' spoolin' spoolin'",	    	/* descrip */
	0	    	    	    	    	/* argDescrip */
    },
    POPT_TABLEEND
};

static const struct poptOption options[] =
{
    {
    	"string",	    	    	    	/* longname */
	's',  	    	    	    	    	/* shortname */
	POPT_ARG_STRING,  	    	    	/* argInfo */
	&stringopt,     	    	    	/* arg */
	0,  	    	    	    	    	/* val 0=don't return */
	"how long is a piece of string",    	/* descrip */
	0	    	    	    	    	/* argDescrip */
    },
    {
    	"bool",	    	    	    	    	/* longname */
	'b',  	    	    	    	    	/* shortname */
	POPT_ARG_NONE,  	    	    	/* argInfo */
	&boolopt,     	    	    	    	/* arg */
	0,  	    	    	    	    	/* val 0=don't return */
	"to be or not to be",	    	    	/* descrip */
	0	    	    	    	    	/* argDescrip */
    },
    {
    	"whatever",				/* longname */
	'?',  	    	    	    	    	/* shortname */
	POPT_ARG_INCLUDE_TABLE,			/* argInfo */
	(void *)included_options,		/* arg */
	0,  	    	    	    	    	/* val 0=don't return */
	0,					/* descrip */
	0	    	    	    	    	/* argDescrip */
    },
    POPT_TABLEEND
};

int
main(int argc, char **argv)
{
    int rc;
    poptContext con;
    const char *file;
    GList *iter;
    const char *argv0;

    argv0 = strrchr(argv[0], '/');
    if (argv0 == NULL)
	argv0 = argv[0];
    else
	argv0++;

    con = poptGetContext("popttest", argc, (const char**)argv, options, 0);
    poptSetOtherOptionHelp(con,
    	    	           "[OPTIONS] [executable|source|directory]...");

    while ((rc = poptGetNextOpt(con)) > 0)
    	;
    if (rc < -1)
    {
    	fprintf(stderr, "%s:%s at or near %s\n",
	    argv0,
	    poptStrerror(rc),
	    poptBadOption(con, POPT_BADOPTION_NOALIAS));
    	exit(0);
    }
    
    while ((file = poptGetArg(con)) != 0)
	files = g_list_append(files, (gpointer)file);
	
    poptFreeContext(con);

    printf("boolopt=%d\n", boolopt);
    printf("spoolopt=%d\n", spoolopt);
    printf("stringopt=\"%s\"\n", stringopt);
    printf("ropeopt=\"%s\"\n", ropeopt);
    printf("files=");
    for (iter = files ; iter != 0 ; iter = iter->next)
    	printf(" \"%s\"", (const char *)iter->data);
    printf("\n");
    return 0;
}
