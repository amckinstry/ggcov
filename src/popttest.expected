===src/popttest 
boolopt=0
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files=
===src/popttest -b
boolopt=1
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files=
===src/popttest --bool
boolopt=1
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files=
===src/popttest -b -b
boolopt=1
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files=
===src/popttest -b -S
boolopt=1
spoolopt=1
stringopt="before"
ropeopt="BEFORE"
files=
===src/popttest --bool --spool
boolopt=1
spoolopt=1
stringopt="before"
ropeopt="BEFORE"
files=
===src/popttest -bS
boolopt=1
spoolopt=1
stringopt="before"
ropeopt="BEFORE"
files=
===src/popttest -s
src/popttest:no value at or near -s
===src/popttest -s FOO
boolopt=0
spoolopt=0
stringopt="FOO"
ropeopt="BEFORE"
files=
===src/popttest --string FOO
boolopt=0
spoolopt=0
stringopt="FOO"
ropeopt="BEFORE"
files=
===src/popttest --string=FOO
boolopt=0
spoolopt=0
stringopt="FOO"
ropeopt="BEFORE"
files=
===src/popttest -b -S -s FOO
boolopt=1
spoolopt=1
stringopt="FOO"
ropeopt="BEFORE"
files=
===src/popttest -bS -s FOO
boolopt=1
spoolopt=1
stringopt="FOO"
ropeopt="BEFORE"
files=
===src/popttest -bSSb -s FOO
boolopt=1
spoolopt=1
stringopt="FOO"
ropeopt="BEFORE"
files=
===src/popttest -bSSb -s FOO -R BAR
boolopt=1
spoolopt=1
stringopt="FOO"
ropeopt="BAR"
files=
===src/popttest -bSSb -sR FOO BAR
boolopt=1
spoolopt=1
stringopt="FOO"
ropeopt="BAR"
files=
===src/popttest --string FOO --rope BAR
boolopt=0
spoolopt=0
stringopt="FOO"
ropeopt="BAR"
files=
===src/popttest --fmeh
src/popttest:bad option at or near --fmeh
===src/popttest -z
src/popttest:bad option at or near -z
===src/popttest -R
src/popttest:no value at or near -R
===src/popttest foo
boolopt=0
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files= "foo"
===src/popttest foo bar
boolopt=0
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files= "foo" "bar"
===src/popttest -b foo bar
boolopt=1
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files= "foo" "bar"
===src/popttest -b -- foo
boolopt=1
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files= "foo"
===src/popttest -b -- -R foo
boolopt=1
spoolopt=0
stringopt="before"
ropeopt="BEFORE"
files= "-R" "foo"
